data/prepare_dataset_giant_midi.py

```diff
@@ -0,0 +1,122 @@
+import os
+import math
+
+import fortepyan as ff
+from tqdm import tqdm
+from datasets import Value, Dataset, Features, Sequence, DatasetDict, load_dataset
+
+from data.quantizer import MidiQuantizer
+
+
+def process_record(piece: ff.MidiPiece, sequence_len: int, quantizer: MidiQuantizer) -> list[dict]:
+    piece_quantized = quantizer.quantize_piece(piece)
+
+    midi_filename = piece_quantized.source["midi_filename"]
+
+    piece.df["next_start"] = piece.df.start.shift(-1)
+    piece.df["dstart"] = piece.df.next_start - piece.df.start
+    piece.df["dstart"].fillna(0, inplace=True)
+
+    record = []
+
+    for subset, subset_quantized in zip(
+        piece.df.rolling(window=sequence_len, step=sequence_len),
+        piece_quantized.df.rolling(window=sequence_len, step=sequence_len),
+    ):
+        # rolling sometimes creates subsets with shorter sequence length, they are filtered here
+        if len(subset) != sequence_len:
+            continue
+
+        sequence = {
+            "midi_filename": midi_filename,
+            "pitch": subset.pitch.astype("int16").values.T,
+            "dstart": subset.dstart.astype("float32"),
+            "duration": subset.duration.astype("float32"),
+            "velocity": subset.velocity.astype("int16"),
+            "dstart_bin": subset_quantized.dstart_bin.astype("int8").values.T,
+            "duration_bin": subset_quantized.duration_bin.astype("int8").values.T,
+            "velocity_bin": subset_quantized.velocity_bin.astype("int8").values.T,
+        }
```

Comment:
```suggestion
        sequence = {
            "midi_filename": midi_filename,
            "pitch": subset.pitch.astype("int16").values.T,
            "dstart": subset.dstart.astype("float32"),
            "duration": subset.duration.astype("float32"),
            "velocity": subset.velocity.astype("int16"),
            "dstart_bin": subset_quantized.dstart_bin.astype("int8").values.T,
            "duration_bin": subset_quantized.duration_bin.astype("int8").values.T,
            "velocity_bin": subset_quantized.velocity_bin.astype("int8").values.T,
            "source": json.dumps(part.source),
        }
```

Let's track the source